src/librbd/api/Mirror.cc

```diff
@@ -415,18 +429,205 @@ struct C_ImageSnapshotCreate : public Context {
 
     auto req = mirror::snapshot::CreatePrimaryRequest<I>::create(
       ictx, mirror_image.global_image_id, CEPH_NOSNAP, snap_create_flags, 0U,
-      snap_id, on_finish);
+      group_pool_id, group_id, group_snap_id, snap_id, on_finish);
     req->send();
   }
 };
 
+int group_image_list_by_id(librados::IoCtx& group_ioctx,
+                           const std::string &group_id,
+                           std::vector<cls::rbd::GroupImageStatus> *image_ids) {
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+
+  std::string group_header_oid = librbd::util::group_header_name(group_id);
+
+  ldout(cct, 20) << "listing images in group id " << group_header_oid << dendl;
+  image_ids->clear();
+
+  int r = 0;
+  const int max_read = 1024;
+  cls::rbd::GroupImageSpec start_last;
+  do {
+    std::vector<cls::rbd::GroupImageStatus> image_ids_page;
+
+    r = cls_client::group_image_list(&group_ioctx, group_header_oid,
+                                     start_last, max_read, &image_ids_page);
+
+    if (r < 0) {
+      lderr(cct) << "error reading image list from group: "
+	<< cpp_strerror(-r) << dendl;
+      return r;
+    }
+    image_ids->insert(image_ids->end(),
+		     image_ids_page.begin(), image_ids_page.end());
+
+    if (image_ids_page.size() > 0)
+      start_last = image_ids_page.rbegin()->spec;
+
+    r = image_ids_page.size();
+  } while (r == max_read);
+
+  return 0;
+}
+
+int group_remove_image(librados::IoCtx& group_ioctx, std::string group_id,
```

Comment:
This does not appear to be doing anything mirror realated. Should it be moved to Group.cc?